src/programmatic_policy_learning/data/demo_types.py

```diff
@@ -0,0 +1,22 @@
+"""Types for demonstrations and trajectories."""
+
+from dataclasses import dataclass
+from typing import Generic, Sequence, TypeVar
+
+ObsT = TypeVar("ObsT")
+ActT = TypeVar("ActT")
+
+
+@dataclass(frozen=True)
+class Demo(Generic[ObsT, ActT]):
```

Comment:
in my opinion, this is one data structure too many -- we can probably get away with just `Trajectory` and, for example, include `obs: list[Obs]` and `act: list[Act]` as fields. if you want to be careful, you can do something like
```python
def __post_init__(self):
    assert len(self.obs) == len(self.act)
```

if you want to keep these two data structures, let's change the name of this one because "Demo" to me means a full trajectory